[Attention] Use head_dim_v as gemm1MPerBlock where possible

manupak opened the pull request:
Builds on https://github.com/ROCm/rocMLIR/pull/1428. Hence that one needs merging first.

This PR makes the gemm1 tuning param derivation always use head dimension of V as gemm1MPerBlock  so that second gemm will not be split across multiple workgroups.

Additionally, it adds a safe guard to not blow up if head_dim_v becomes too large resulting whole tuning space becoming invalid.

closes : https://github.com/ROCm/rocMLIR-internal/issues/1392

manupak committed:
* define attentionV2 with a region that describes a fusion point
* handle the tosa-to-rock lowering of that

manupak committed:
* folded the changes back to attention op (removed v2)
* end to end kernel pipeline is producing something
* need to handle the region in gridwise lowering

manupak committed:
* changed gridwise gemm to blockwise gemm to align the linalg
  block in preSoftmax fusion point.
* TODO : add support for unary elementwise in tosa to rock.

manupak committed:
* add view like ops into the mix
* cleaned up tosa to rock
* added handling of unary ops
* TODO: untransfrom + invert la generic inputs to tileviews

manupak committed:
* added handling of reshapes inside elemwise tree
* added more tests to cover elemwise fusion cases

manupak committed:
Merge branch 'develop' of https://github.com/ROCmSoftwarePlatform/rocMLIR into attn-presoftmax-fusion-point

manupak committed:
* remove some deadcode

manupak committed:
* clang format

manupak committed:
* remove commented code

manupak committed:
* add gridwise attention op verifier to check #linalg ops

manupak committed:
* clang format

manupak committed:
* non-padded first gemm i8 working

manupak committed:
* unpad the transforms before elemwise fusion
* add a padded i8 migraphx e2e test

manupak committed:
* add i8-f16 migraphx test

manupak committed:
* added rocmlir-gen support for i8 attention (first gemm only)
* added tests for i8

manupak committed:
* clang format

manupak committed:
Merge branch 'develop' of https://github.com/ROCmSoftwarePlatform/rocMLIR into attn-i8-first-gemm

manupak committed:
Remove padding from gemm to gridwise for other elemwise
inputs as now we take gemm0 out view to be unpadded prior
to linalg fusion

manupak committed:
* clang format

manupak committed:
* clean postProcessFirstGemm

manupak committed:
further clean up

manupak committed:
Merge branch 'develop' of https://github.com/ROCmSoftwarePlatform/rocMLIR into attn-i8-first-gemm

manupak committed:
* enable fusion of tosa.const into the region

manupak committed:
Merge branch 'develop' of https://github.com/ROCmSoftwarePlatform/rocMLIR into attn-i8-first-gemm

manupak committed:
* dont pass the region if it do not have
  a linalg generic op in it.

manupak committed:
Changed non test code to new problem key

manupak committed:
* change self attention problem keys to generalized attention problem
  key

manupak committed:
* clang format

manupak committed:
Merge branch 'develop' of https://github.com/ROCmSoftwarePlatform/rocMLIR into attn-cross-problem-key

manupak committed:
* revert removal of nighlty-only enable

manupak committed:
make gemm1MPerBlock to be gemm1M

manupak committed:
* change gemm1MPerBlock to be decided by tuning params

manupak committed:
* make gemm1Mperblock be gemm1M where possible

manupak committed:
* clang format

manupak committed:
* fix affix tuning params test

krzysz00 reviewed (approved):
This all feels reasonable to me

manupak committed:
Workaround nan failure when MPerWave < 32

This commit adds checks to see if the derived
perfconfig is valid. Additionaly, there was
a nan failure when gemm1MperBlock is increased
to gemm1M when MperWave of gemm0 is < 32. This
commit disables the optimization when that happens.

manupak committed:
* only perform the optmization when its needed

manupak committed:
* clang format
